pandas/core/algorithms.py

```diff
@@ -948,10 +948,39 @@ def select_n_frame(frame, columns, n, method, keep):
     if not is_list_like(columns):
         columns = [columns]
     columns = list(columns)
-    ser = getattr(frame[columns[0]], method)(n, keep=keep)
-    if isinstance(ser, Series):
-        ser = ser.to_frame()
-    return ser.merge(frame, on=columns[0], left_index=True)[frame.columns]
+    ascending = method == 'nsmallest'
+    original_index = frame.index
+    frame = frame.reset_index(drop=True)
+    for i, column in enumerate(columns):
+
+        # For each column in columns we peform ``method`` on this frame
+        # To guard against the possibility column has duplicate values that
+        # must be considered for futher columns (# GH15297) we filter using
+        # frame[isin] on the values returned by ``method``. If there are no
+        # duplicated values, we simply reindex on the values returned by
+        # ``method``, otherwise we sort the isin filtered frame and continue
+        series = frame[column]
+        values = getattr(series, method)(n, keep=keep)
+        if i + 1 == len(columns):
```

Comment:
```
values = .....
indexer = values.index
```

Then just ``frame.take(indexer)``

its obvious then this is positional taking (rather than ``.reindex``) that this *might* be label based.